hadoop-hdfs-project/hadoop-hdfs/src/main/java/io/hops/metadata/security/token/block/NameNodeBlockTokenSecretManager.java

```diff
@@ -53,151 +49,49 @@ public NameNodeBlockTokenSecretManager(long keyUpdateInterval,
       return updateBlockKeys();
     } else {
-      retrieveBlockKeys();
+      getAllKeysAndSync();
+      removeExpiredKeys();
     }
     return true;
   }
 
-  @Override
-  public DataEncryptionKey generateDataEncryptionKey() throws IOException {
-    byte[] nonce = new byte[8];
-    nonceGenerator.nextBytes(nonce);
-    BlockKey key = getBlockKeyByType(BlockKey.KeyType.CurrKey);
-
-    byte[] encryptionKey = createPassword(nonce, key.getKey());
-    return new DataEncryptionKey(key.getKeyId(), blockPoolId, nonce,
-        encryptionKey, Time.now() + tokenLifetime, encryptionAlgorithm);
-  }
-
-  @Override
-  protected byte[] createPassword(BlockTokenIdentifier identifier) {
-    BlockKey key;
-    try {
-      key = getBlockKeyByType(BlockKey.KeyType.CurrKey);
-    } catch (IOException ex) {
-      throw new IllegalStateException(
-          "currentKey hasn't been initialized. [" + ex.getMessage() + "]");
-    }
-    if (key == null) {
-      throw new IllegalStateException("currentKey hasn't been initialized.");
-    }
-    identifier.setExpiryDate(Time.now() + tokenLifetime);
-    identifier.setKeyId(key.getKeyId());
-    if (LOG.isDebugEnabled()) {
-      LOG.debug("Generating block token for " + identifier.toString());
-    }
-    return createPassword(identifier.getBytes(), key.getKey());
-  }
-
-  @Override
-  public byte[] retrievePassword(BlockTokenIdentifier identifier)
-      throws InvalidToken {
-    if (isExpired(identifier.getExpiryDate())) {
-      throw new InvalidToken(
-          "Block token with " + identifier.toString() + " is expired.");
-    }
-    BlockKey key = null;
-    try {
-      key = getBlockKeyById(identifier.getKeyId());
-    } catch (IOException ex) {
-    }
-
-    if (key == null) {
-      throw new InvalidToken(
-          "Can't re-compute password for " + identifier.toString() +
-              ", since the required block key (keyID=" + identifier.getKeyId() +
-              ") doesn't exist.");
-    }
-    return createPassword(identifier.getBytes(), key.getKey());
-  }
-
-  public void generateKeysIfNeeded() throws IOException {
-    if (isLeader()) {
-      retrieveBlockKeys();
-      if (currentKey == null && nextKey == null) {
-        generateKeys();
+  protected boolean updateKeysParent() throws IOException {
+    return super.updateKeys();
+  }
+
+  public void initKeys() throws IOException {
+    while (true) {
+      if (isLeader()) {
+        /*
+         * If the NN is leader at this point it means that the all cluster was restarted
+         * the NN should then write its newly generated keys to the database
+         */
+        addBlockKeys();
+        return;
+      } else {
+        /*
+         * if the NN is not the leader it try to get keys until the leader as persisted keys up to date neough to
+         * the database
+         */
+        getAllKeysAndSync();
+        if (currentKey.getExpiryDate() > Time.now() + 2 * keyUpdateInterval + tokenLifetime) {
```

Comment:
If NameNode is not a leader and gets into this loop, should we add a timeout mechanism? If the leader failed to add keys for whatever reason, Slaves should not block here. I think it's better to throw an exception and shut them down. On the other hand, a Slave might be promoted to Leader and get into the `if(isLeader())` path, so the timeout must be long enough...